File: server/src/main/java/dev/slimevr/posestreamer/PoseFrameStreamer.java
package dev.slimevr.posestreamer;

import dev.slimevr.config.ConfigManager;
import dev.slimevr.poserecorder.PoseFrameIO;
import dev.slimevr.poserecorder.PoseFrames;
import dev.slimevr.tracking.processor.HumanPoseManager;

import java.io.File;


public class PoseFrameStreamer extends PoseStreamer {

	private final PoseFrames frames;
	private final HumanPoseManager humanPoseManager;

	public PoseFrameStreamer(String path) {
		this(new File(path));
	}

	public PoseFrameStreamer(File file) {
		this(PoseFrameIO.readFromFile(file));
	}

	public PoseFrameStreamer(PoseFrames frames) {
		this.frames = new PoseFrames(frames);
		humanPoseManager = new HumanPoseManager(this.frames.getTrackers());
		skeleton = humanPoseManager.getSkeleton();
	}

	public PoseFrameStreamer(PoseFrames frames, ConfigManager configManager) {
		this(frames);
		humanPoseManager.loadFromConfig(configManager);
	}

	public PoseFrames getFrames() {
		return frames;
	}

	public HumanPoseManager getHumanPoseManager() {
		return humanPoseManager;
	}

	public synchronized void streamAllFrames() {
		for (int i = 0; i < frames.getMaxFrameCount(); i++) {
			frames.setCursors(i);
			humanPoseManager.update();
			captureFrame();
		}
	}
}

File: server/core/src/main/java/dev/slimevr/posestreamer/PoseFrameStreamer.kt
package dev.slimevr.posestreamer

import dev.slimevr.poseframeformat.PfrIO.readFromFile
import dev.slimevr.poseframeformat.PoseFrames
import dev.slimevr.poseframeformat.player.TrackerFramesPlayer
import dev.slimevr.tracking.processor.HumanPoseManager
import java.io.File

class PoseFrameStreamer : PoseStreamer {

	val player: TrackerFramesPlayer
	val hpm: HumanPoseManager

	private constructor(
		player: TrackerFramesPlayer,
		hpm: HumanPoseManager,
	) : super(hpm.skeleton) {
		this.player = player
		this.hpm = hpm
	}

	constructor(player: TrackerFramesPlayer) : this(player, HumanPoseManager(player.trackers.toList()))
	constructor(poseFrames: PoseFrames) : this(TrackerFramesPlayer(poseFrames))
	constructor(file: File) : this(readFromFile(file))
	constructor(path: String) : this(File(path))

	@Synchronized
	fun streamAllFrames() {
		for (i in 0 until player.maxFrameCount) {
			player.setCursors(i)
			hpm.update()
			captureFrame()
		}
	}
}
